Fix get(x::Nullable{Union{}}, y) and add tests for Nullable{Union{}}

nalimilan opened the pull request:
It failed since conversion of y to Union{} wasn't possible. Add tests
for Nullable() everywhere the behavior of a null is tested.

This generates type stable code according to `@code_warntype` (except of course when the promotion rules give `Any`).

Cc: @TotalVerb @johnmyleswhite

nalimilan committed:
Fix get(x::Nullable{Union{}}, y) and add tests for Nullable{Union{}}

It failed since conversion of y to Union{} wasn't possible. Stop
converting y for consistency with other get() methods.
Add tests for Nullable() everywhere the behavior of a null is tested.

vtjnash commented:
I'm not sure that calling `promote_type` is quite right either, or would just causes the same sort of problems in the opposite direction.

For the other container types in base, `default` is returned exactly (without trying to convert it to anything else). I think we could do the same here.

Alternatively, if we don't want to change the (current, documented) behavior of `get`, than I think it would be more appropriate to add a special-case method for `Union{}`: `get(::Nullable{Union{}}, default) = default` since the `Nullable{Union{}}` type is already known to be treated as a special-case sentinel in some cases (such as `Nullable()`).

nalimilan commented:
> I'm not sure that calling promote_type is quite right either, or would just causes the same sort of problems in the opposite direction.

What kind of problems do you expect?

nalimilan commented:
I've updated the PR to stop converting the argument for consistency with other `get` methods. This is indeed simpler, though we could imagine changing all of them if we find cases in which it would be useful.

I've also tried to improve docs a bit and moved them inline.

nalimilan commented:
If nobody has other remarks I'll merge tomorrow.

johnmyleswhite commented on base/nullable.jl:
```diff
@@ -45,16 +45,22 @@ function show{T}(io::IO, x::Nullable{T})
+"""
+@inline function get{S,T}(x::Nullable{S}, y::T)
+    if isbits(S)
+        ifelse(x.isnull, y, x.value)
```
Given your recent work on `null_safe_op`, should you revisit this line?

nalimilan replied:
Not sure. For `isbits` types, it should always be safe to eagerly evaluate `x.value`, even if it's nonsense. Maybe we don't need `null_safe_op` here (what's "op" BTW?).

omus commented:
This PR broke a doctest:
```julia
julia> get(Nullable{Float64}(), 0)
0.0
```
It makes sense to have the `get` method to be consistent. I'll update the doctest.
